## EDA: Null values

In [1]:
import pandas as pd

### reading our data (.csv file) into dataframe

In [5]:
scraped_data_bt = pd.read_csv('scraped_csv_one.csv',parse_dates=True, engine='python')
scraped_data_bt.head()

content   location        date
0  Bt are awful - terrible communication. I wante...     TOTNES  2021-11-22
1  Very poor service inspite of being with BT for...       None        None
2                Absolutely rubbish customer service  wakefield  2021-11-19
3  Shambles of a company. I couldn`t use the Broa...  Stockport  2021-11-14
4  The service they provide is not close to that ...  Kirkcaldy  2021-11-14

### Lets drop rows with null values in all columns

(we observe that there are no rows that contains null value in all columns)

In [3]:
scraped_data_bt.dropna(axis=0, how='all')

content           location  \
0     Bt are awful - terrible communication. I wante...             TOTNES   
1     Very poor service inspite of being with BT for...               None   
2                   Absolutely rubbish customer service          wakefield   
3     Shambles of a company. I couldn`t use the Broa...          Stockport   
4     The service they provide is not close to that ...          Kirkcaldy   
...                                                 ...                ...   
6026  Almost impossible to contact BT when problems ...               Leek   
6027  Nothing but trouble - needed new land line ins...  Egremont, Cumbria   
6028  I use to get Just over 6 at a previous address...   Barnstaple Devon   
6029  Shocked and disappointed by the results of the...             Rosyth   
6030  I really though I had bought a fast service bu...             London   

            date  
0     2021-11-22  
1           None  
2     2021-11-19  
3     2021-11-14  
4     2021-11-14  
...          ...  
6026  2009-07-29  
6027  2009-07-28  
6028  2009-07-28  
6029  2009-07-28  
6030  2009-07-28  

[6031 rows x 3 columns]

### Lets check for duplicate values in our dataframe

In [4]:
scraped_data_bt[scraped_data_bt.duplicated()].count()

content     118
location     54
date         54
dtype: int64

<div class="alert alert-block alert-warning">
On furthur investigation, we found 2 reasons for duplications.
    
1. Duplicates exist in the data source (https://www.broadband.co.uk/broadband/providers/bt/#reviewers)
2. Our web scraping code was duplicating the data from pages 1-4. 
</div>

<div class="alert alert-block alert-info">
Actions taken:
    
1. Duplicates that exists in the data source is deleted.
2. web scraping code is rectified.
</div>

In [6]:
bt_data = scraped_data_bt.drop_duplicates()
bt_data

content           location  \
0     Bt are awful - terrible communication. I wante...             TOTNES   
1     Very poor service inspite of being with BT for...               None   
2                   Absolutely rubbish customer service          wakefield   
3     Shambles of a company. I couldn`t use the Broa...          Stockport   
4     The service they provide is not close to that ...          Kirkcaldy   
...                                                 ...                ...   
6026  Almost impossible to contact BT when problems ...               Leek   
6027  Nothing but trouble - needed new land line ins...  Egremont, Cumbria   
6028  I use to get Just over 6 at a previous address...   Barnstaple Devon   
6029  Shocked and disappointed by the results of the...             Rosyth   
6030  I really though I had bought a fast service bu...             London   

            date  
0     2021-11-22  
1           None  
2     2021-11-19  
3     2021-11-14  
4     2021-11-14  
...          ...  
6026  2009-07-29  
6027  2009-07-28  
6028  2009-07-28  
6029  2009-07-28  
6030  2009-07-28  

[5913 rows x 3 columns]

### Lets check for null values in our dataframe

In [7]:
bt_data.isnull().sum()

content        3
location    2796
date        2739
dtype: int64

<div class="alert alert-block alert-warning">
Observation/Reasons for null values.
    
1. Content column has 3 rows of null which is observed in the data source as well.
2. Location and date columns has large number of null values. It was observed that the code for scraping data had a bug which was splitting reviews based on paragraphs (or \n) and null values were getting inserted in the location and date columns (2739 rows). 
3. The remaining 57 values of null values in the location column was missing from our data souce as well.
    (https://www.broadband.co.uk/broadband/providers/bt/#reviewers)
</div>

<div class="alert alert-block alert-info">
Actions taken:
    
1. web scraping code is rectified to avoid inserting null values in the location and date column.
2. Remaining 57 null values in location column is replaced with 'UK' since all our reviews are based from UK.
</div>

In [9]:
bool_series_loc = pd.isnull(bt_data['location'])
bool_series_date = pd.isnull(bt_data['date'])

replace_null_loc = bt_data[bool_series_loc & ~bool_series_date]
replace_null_loc['location'].fillna('UK', inplace = True)
replace_null_loc

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


content location        date
473   Terrible on all levels. Had a multiple problem...       UK  2020-10-24
580   Utterly abysmal! Constant disconections  never...       UK  2020-05-13
643   Cant even open a web page even when no one els...       UK  2020-01-25
721   On top of that the complaints procedure is obv...       UK  2019-10-06
829    This didn't happen. Weeks and many phone call...       UK  2019-06-07
869   I do not recommend BT they promised me a £39.9...       UK  2019-04-27
1014  We have the Infinity package which averages 0....       UK  2018-12-21
1043  Following resetting of an e-mail password by a...       UK  2018-12-03
1136  BT business i would never recommend this compa...       UK  2018-10-03
1143  Every 5 minutes the internet would fail and ki...       UK  2018-09-30
1216   Avoid if you're a student or  for any other r...       UK  2018-09-07
1460  Honestly feels like a time machine. Reminds me...       UK  2018-06-07
1560  I had BT for a month I received first bill wit...       UK  2018-04-16
1609  I would never recommend you to anybody. If I'm...       UK  2018-03-30
1625  Most expensive internet provider in the countr...       UK  2018-03-24
1679  DO NOT hang up the phone without getting a con...       UK  2018-03-08
1712  We moved house 2 weeks ago  was meant to have ...       UK  2018-02-23
1729  I would not give even 1 star. Horrible custome...       UK  2018-02-15
1743   I was quite annoyed with this because when th...       UK  2018-02-07
1845  Called BT to enquire about installing broadban...       UK  2017-12-28
2042  It's rubbish and a waste of money. Took a minu...       UK  2017-10-01
2117  Terrible service once again from BT! We were j...       UK  2017-09-17
2138  absolutely appalling broadband. go against the...       UK  2017-09-12
2204  Slow  poor support. Never listen and their spe...       UK  2017-08-20
2248  Such a bad internet connection  the devices sa...       UK  2017-08-08
2395   The broadband speed is ok but everything else...       UK  2017-06-23
2522  This is the worst company I have ever dealt wi...       UK  2017-05-12
2710  never purchase bt because you never get the sp...       UK  2017-03-30
2768  Absolutely terrible !!  staff are so rude. Hun...       UK  2017-03-14
2924  Worst broadband ever  plus that expensive... i...       UK  2017-01-25
3064  Horrifyingly slow speeds with bt infinity!! Mu...       UK  2016-12-09
3277  Never even got to use the service. They took 5...       UK  2016-10-24
3336  We had no contact prior to the contract ending...       UK  2016-10-15
3418  Absolutely useless!!!!!! Took 6 weeks to activ...       UK  2016-09-27
3555  DON'T GET SUCKERED INTO ONE OF BT'S DEALS! I s...       UK  2016-08-20
3609  Used to have speed of 1.4 mb year ago. Now luc...       UK  2016-08-10
3724  I wish there was a zero star for rating becaus...       UK  2016-07-18
3743   In my last call (note not final  but last) th...       UK  2016-07-12
4006  Worst  sorry excuse of a company I have ever h...       UK  2016-04-07
4064  Random disconnects happen far too often  even ...       UK  2016-03-28
4081  BT has the worst service I ever had in my life...       UK  2016-03-23
4172  Bt is a bunch of layers. After I order my broa...       UK  2016-02-26
4231  O Stars   I have been around the world on the ...       UK  2016-02-06
4384  Couldn't be unhappier with BT. The speed is ok...       UK  2015-12-12
4576                    stay away you are warned!!!!!!!       UK  2015-09-30
4609  I'm incredibly disappointed. Especially that I...       UK  2015-09-21
4646  I signed up to BT on the 18th August. After 2 ...       UK  2015-09-09
4827  Awful customer service from the call centre!! ...       UK  2015-07-14
4838  Absolutely disgusted at the dishonesty & fraud...       UK  2015-07-09
4924  Sick to the back teeth of BT and their NOT Sup...       UK  2015-06-14
5062  i could go on for ever about BT bad service bu...       UK  2015-04-13
5076  DO NOT use BT if you suffer with any f

In [10]:
# to merge replaced values with original dataset
# orig_df.combine_first(new_df)

bt_data = bt_data.combine_first(replace_null_loc)
bt_data

content           location  \
0     Bt are awful - terrible communication. I wante...             TOTNES   
1     Very poor service inspite of being with BT for...                NaN   
2                   Absolutely rubbish customer service          wakefield   
3     Shambles of a company. I couldn`t use the Broa...          Stockport   
4     The service they provide is not close to that ...          Kirkcaldy   
...                                                 ...                ...   
6026  Almost impossible to contact BT when problems ...               Leek   
6027  Nothing but trouble - needed new land line ins...  Egremont, Cumbria   
6028  I use to get Just over 6 at a previous address...   Barnstaple Devon   
6029  Shocked and disappointed by the results of the...             Rosyth   
6030  I really though I had bought a fast service bu...             London   

            date  
0     2021-11-22  
1            NaN  
2     2021-11-19  
3     2021-11-14  
4     2021-11-14  
...          ...  
6026  2009-07-29  
6027  2009-07-28  
6028  2009-07-28  
6029  2009-07-28  
6030  2009-07-28  

[5913 rows x 3 columns]